rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,275 @@
+  - Written in Rust:
+    - [Tantivy](https://github.com/tantivy-search/tantivy)
+    - [Toshi](https://github.com/toshi-search/Toshi)
+    - [Sonic](https://github.com/valeriansaliou/sonic)
+    - [MeiliSearch](https://github.com/meilisearch/MeiliSearch)
+    - [Bayard](https://github.com/bayard-search/bayard)
+
+## Compatibility
+
+This proposal does not change any existing interfaces, so no migrations 
+will be necessary for existing subgraph deployments. 
+
+## Drawbacks and Risks
+
+There is little risk in implementing this feature.  The proposed solution 
+uses native Postgres fulltext features and there is a nonzero probability 
+this choice results in slower than optimal write and read times; 
+however the tradeoff in implementation time/complexity and the existence 
+of production use case testimonials tempers my apprehension here. 
+
+Are there any desired "fulltext search like" capabilities that I've missed in this proposal?   
+
+## Alternatives
+
+An alternative design for the feature would allow more flexibility for 
+Graph Node operators in their index implementation and create a marketplace 
+for indexes. This design moved the definition of custom indexes out of the 
+subgraph definition; the subgraph would be deployed without them and they 
+could be added later using a new Graph Explorer interface (in Hosted-Service context) 
+or a JSON-RPC request directly to a Graph Node. One of the benefits of this 
+design is that the definition of uniqueness for a subgraph does not include 
+the custom indexes, so a new subgraph deployment and corresponding syncing 
+work does not have to be added in order to create or update an index. 
+However, it also introduces significant added complexity: will a separate 
+query marketplace and discovery registry then be required for finding nodes 
+with your needed subgraph-index combination.  
+
+## Open Questions
+
+Full-text search queries introduce new issues with maintaining query 
+result determinism which will become a more potent issue with the 
+decentralized network. A fulltext search query and a dataset are not enough 
+to determine the output of the query, the index is vital to establish a 
+deterministic causal relationship to the output data.  Query verification 
+will need to take into account the query, the index, the underlying dataset, 
+and the query result.  Can we find a healthy compromise between being 
+prescriptive about the indexes and algorithms in order to allow formal 
+verification and allowing indexer node operators to experiment with 
+algorithms and indexes in order to continue to improve query speed and results? 
+
+Since a fulltext search field is purely derivative of other Entity data 
+the addition or update of an @index field does not require a full blockchain 
+resync, rather the index itself just needs to be rebuilt. 
+How can we allow fulltext search index (and eventually other index types) 
+updates without requiring a full subgraph resync? 
```

Comment:
We can't currently. For the moment, I feel that's ok. Should we rename `@index field` to `@fulltext directive`?